c++/src/kj/async-inl.h

```diff
@@ -306,18 +314,97 @@ class PromiseNode: private AsyncObject {
   private:
     Event* event = nullptr;
   };
+
+private:
+  PromiseArena* arena = nullptr;
+  // If non-null, then this PromiseNode is the last node allocated within the given arena, and
+  // therefore owns the arena. After this node is destroyed, the arena should be deleted.
+  //
+  // PromiseNodes are allocated within the arena starting from the end, and `PromiseNode`s
+  // allocated this way are required to have `PromiseNode` itself as their leftmost inherited type,
+  // so that the pointers match. Thus, the space in `arena` from its start to the location of the
+  // `PromiseNode` is known to be available for subsequent allocations (which should then take
+  // ownership of the arena).
+
+  friend class PromiseDisposer;
 };
 
 class PromiseDisposer {
 public:
   static void dispose(PromiseNode* node) {
+    PromiseArena* arena = node->arena;
     node->destroy();
+    delete arena;
```

Comment:
Maybe worth a comment that `delete` is safe to call on `nullptr`? I had to look it up to be sure.